### PR TITLE
UCP/RNDV: fix rtr/ppl error flow

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
fix rtr/ppl error flow

## Why ?
 - `UCP_REQUEST_FLAG_RNDV_FRAG` is a recv request, so we can not access send part of it
 - leaked `req->send.rndv.mdesc` from rtr request

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
@pentschev FYI

**Comment by evgeny-leksikov** [comment]
fixes https://github.com/openucx/ucx/issues/8669

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -3496,14 +3496,21 @@ void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
...
-        /* SW RMA/PUT and AMO/Post operations don't allocate local request ID
-         * and don't need to be tracked, since they complete UCP request upon
-         * sending all data to a peer. Receiving RMA/CMPL and AMO/REP packets
-         * complete flush requests */
-        ucs_assert((req->send.uct.func != ucp_rma_sw_proto.progress_put) &&
-                   (req->send.uct.func != ucp_amo_sw_proto.progress_post));
         ucs_assert(req->send.ep == ucp_ep);
 
+        if (req->send.uct.func == ucp_proto_progress_rndv_rtr) {
+            if (req->send.rndv.mdesc != NULL) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:3502]: it seems that ucp_rndv_receive->ucp_rndv_req_send_rtr does not initialize desc to NULL

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:3502]: yeah, it wrongly set by a caller
[`        frndv_req->send.rndv.mdesc   = mdesc;`](https://github.com/openucx/ucx/blob/7ea28295471b18f691ec449d39dee97bd11e6553/src/ucp/rndv/rndv.c#L1405)
which uses other union part of send request `send.rndv` but `ucp_proto_progress_rndv_rtr` uses `send.rndv_rtr`

Comment (yosefe) [reply src/ucp/core/ucp_ep.c:3502]: maybe we should check RNDV_FRAG request flag?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:3502]: RTR does not have this flag

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -3473,7 +3474,6 @@ void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
         ucp_request_complete_tag_recv(req, status);
     } else if (req->flags & UCP_REQUEST_FLAG_RNDV_FRAG) {
         ucs_assert(req->flags & UCP_REQUEST_FLAG_SUPER_VALID);
-        ucs_assert(req->send.ep == ucp_ep);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:3476]: why?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:3476]: see PR description, this is `recv` request, we can not access its `send` part

Comment (yosefe) [reply src/ucp/core/ucp_ep.c:3476]: how can it be recv request if we removed it from the pending queue?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:3476]: this is a super request of RTR, hierarchy of completions is 
```
rtr_req -> frag_req -> recv_req
(send)     (recv)      (recv)      
```

Comment (yosefe) [reply src/ucp/core/ucp_ep.c:3476]: i see
https://github.com/openucx/ucx/blob/dfdd4f4dd733d77c38a70f8de4e3366f8d4c060d/src/ucp/rndv/rndv.c#L1378

**Review by yosefe — APPROVED** [review]

**Pull request merged**